File: src/main/java/bllose/sortrelated/KakaEatingBananas.java
package bllose.sortrelated;

import java.util.Arrays;
import java.util.Comparator;

import static java.lang.Math.ceil;

public class KakaEatingBananas {

    public static int KekelovesBanana(Integer[] piles, int k){
        if(piles.length > k){
            return -1;
        }
        Comparator<Integer> comparator = Comparator.comparing(Integer::intValue);
        Integer max = Arrays.stream(piles).max(comparator).get();
        if(piles.length == k){
            return max;
        }

        Integer min = 1;

        while(max != min){
            Integer middle = min + (max - min)/2;
            if(couldEatAllOfBananas(piles, k, middle)){
                if(max==middle) {
                    max --; middle--;
                }else {
                    max = middle;
                }
            }else{
                if(min == middle){
                    min ++;
                    middle ++;
                }else{
                    min = middle;
                }
            }
        }
        return min;
    }

    private static boolean couldEatAllOfBananas(Integer[] piles, int k, Integer middle) {
        int times = 0;
        for(int i = 0 ; i < piles.length; i ++){
            times += ceil((double)piles[i]/middle);
            if(times > k){
                return false;
            }
        }
        return true;
    }
}

File: src/main/python/bllose/sortrelated/KakaEatingBananas.py
from typing import List
from math import ceil

def KakaEatingBananas(pilts:List, times:int) -> int:
    if pilts is None or len(pilts) > times:
        return -1

    right, left = max(pilts), 1

    while left != right:
        middle = int((left + right)/2)
        if eating(pilts, times, middle):
            if right == middle:
                right -= 1
                middle -= 1
            else:
                right = middle
        else:
            if left == middle:
                left += 1
                middle += 1
            else:
                left = middle
    return int(left)


def eating(pilts:List, times:int, eachTime:int) -> bool:
    timeUsed = 0
    for pilt in pilts:
        timeUsed += ceil(pilt/eachTime)
        if timeUsed > times:
            return False
    
    return True




if __name__ == '__main__':
    numEachTime = KakaEatingBananas([1,2,3], 3)
    print(numEachTime)